controllers/spacecompletion/space_completion_controller.go

```diff
@@ -0,0 +1,111 @@
+package spacecompletion
+
+import (
+	"context"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	"github.com/codeready-toolchain/host-operator/controllers/toolchainconfig"
+	"github.com/codeready-toolchain/host-operator/pkg/capacity"
+	"github.com/codeready-toolchain/host-operator/pkg/pending"
+	"github.com/codeready-toolchain/toolchain-common/pkg/cluster"
+	"github.com/go-logr/logr"
+	"github.com/redhat-cop/operator-utils/pkg/util"
+
+	errs "github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Reconciler reconciles a Space object
+type Reconciler struct {
+	Client            client.Client
+	Namespace         string
+	GetMemberClusters cluster.GetMemberClustersFunc
+}
+
+// SetupWithManager sets up the controller reconciler with the Manager
+// Watches the Space resources and the ToolchainStatus CRD
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("spacecompletion").
+		// watch Spaces in the host cluster
+		For(&toolchainv1alpha1.Space{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Watches(
+			&source.Kind{Type: &toolchainv1alpha1.ToolchainStatus{}},
+			handler.EnqueueRequestsFromMapFunc(pending.NewSpaceMapper(mgr.GetClient()).MapToOldestPending)).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=toolchain.dev.openshift.com,resources=spaces,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile ensures that Space has set all missing fields that are needed for proper provisoning
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx, "namespace", r.Namespace)
+	logger.Info("reconciling Space")
+
+	// Fetch the Space
+	space := &toolchainv1alpha1.Space{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: r.Namespace,
+		Name:      request.Name,
+	}, space)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("Space not found")
```

Comment:
the `IsNotFound` case could be tested

Reply:
We don't test these two cases in 99% of all controllers, so I didn't see it important doing that here. 
But I can add test case for that.

Reply:
added c7a09ec8